File: csharp/Vernacula.Phonemizer/Core/RiderDiacritizer.cs
/**
 * Neural GENERALIZATION tier for the Perso-Arabic riders (Urdu, Persian, Pashto, Punjabi-Shahmukhi) — a
 * shared multilingual BiLSTM that restores the short-vowel harakat these abjads leave unwritten, run as an
 * ASYNC pre-pass. ⚠ THE PRECEDENCE IS lexicon → neural → default: a word the exact-match lexicon already
 * covers is LEFT BARE here, so the authoritative sync lexicon layer vocalizes it. Vocalized output feeds
 * the SYNC g2p, so `Phonemize` stays sync.
 * Ported from src/core/riderDiacritizer.ts — see that file for the corpus evidence.
 */
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;

namespace Vernacula.Phonemizer.Core;

public sealed class RiderDiacritizerMeta
{
    public Dictionary<string, int> Chars { get; init; } = new();
    public Dictionary<string, int> Labels { get; init; } = new();
    [JsonPropertyName("lang_tokens")]
    public Dictionary<string, string> LangTokens { get; init; } = new();
}

public interface IRiderDiacritizer
{
    /** Vocalize the words of `text` for `lang`, skipping any word whose skeleton is in `lexicon` (left bare
     *  for the sync lexicon layer). Position-preserving: only harakat are inserted; spacing, digits and
     *  punctuation are kept. */
    Task<string> Diacritize(string text, string lang, IReadOnlyDictionary<string, string> lexicon);
}

public static class RiderDiacritizerLoader
{
    // label → the combining haraka to append after a base letter, mirroring the training VOWELS map. A "~"
    // prefix means shadda (gemination) plus that vowel; "0" is bare, the model's default.
    private static readonly IReadOnlyDictionary<string, string> HAR = new Dictionary<string, string>(StringComparer.Ordinal)
    {
        ["0"] = "", ["a"] = "َ", ["u"] = "ُ", ["i"] = "ِ", ["o"] = "ْ",
        ["F"] = "ً", ["N"] = "ٌ", ["K"] = "ٍ", ["^"] = "ٰ",
    };

    private static string HarOf(string label)
    {
        if (label.StartsWith("~", StringComparison.Ordinal)) return "ّ" + (HAR.GetValueOrDefault(label[1..]) ?? "");
        return HAR.GetValueOrDefault(label) ?? "";
    }

    private static readonly JsRe WORD = JsRegex.Compile("[؀-ۿݐ-ݿ‌‍]+", "gu");

    private sealed class Loaded : IRiderDiacritizer
    {
        private readonly IOrtLike _ort;
        private readonly IOrtSession _session;
        private readonly RiderDiacritizerMeta _meta;
        private readonly string[] _ilabels;
        private readonly int _unk;
        private readonly int _nLabels;

        internal Loaded(IOrtLike ort, IOrtSession session, RiderDiacritizerMeta meta)
        {
            _ort = ort;
            _session = session;
            _meta = meta;
            var ilabels = new string[meta.Labels.Count == 0 ? 0 : meta.Labels.Values.Max() + 1];
            foreach (var (lab, idx) in meta.Labels) ilabels[idx] = lab;
            _ilabels = ilabels;
            _unk = meta.Chars.GetValueOrDefault("<unk>", 1);
            _nLabels = ilabels.Length;
        }

        /**
         * One skeleton → predicted harakat labels (per base char; the prepended lang-token position is
         * dropped).
         */
        private async Task<List<string>> Predict(int langTokIdx, IReadOnlyList<int> chars)
        {
            var seq = new List<long> { langTokIdx };
            seq.AddRange(chars.Select(c => (long)c));
            var input = _ort.Tensor("int64", seq.ToArray(), new[] { 1, seq.Count });
            var outp = await _session.Run(new Dictionary<string, OrtTensor> { ["input"] = input }).ConfigureAwait(false);
            var logitsTensor = outp.GetValueOrDefault("logits") ?? outp.Values.FirstOrDefault();
            if (logitsTensor is null) throw new InvalidOperationException("rider diacritizer ONNX produced no output");
            var logits = logitsTensor.AsFloat32();
            var labels = new List<string>();
            for (var p = 1; p < seq.Count; p++) // p=0 is the lang token → skipped
            {
                var best = 0;
                var bestv = float.NegativeInfinity;
                var @base = p * _nLabels;
                for (var o = 0; o < _nLabels; o++)
                {
                    var v = logits[@base + o];
                    if (v > bestv) { bestv = v; best = o; }
                }
                labels.Add(_ilabels[best]);
            }
            return labels;
        }

        public async Task<string> Diacritize(string text, string lang, IReadOnlyDictionary<string, string> lexicon)
        {
            var langTok = _meta.LangTokens.GetValueOrDefault(lang);
            int? langTokIdx = langTok is not null && _meta.Chars.TryGetValue(langTok, out var li) ? li : null;
            if (langTokIdx is null) return text; // unknown language → no-op
            var @out = new List<string>();
            var cursor = 0;
            foreach (Match m in JsRegex.MatchAll(WORD, text))
            {
                var raw = m.Value;
                @out.Add(text[cursor..m.Index]);
                cursor = m.Index + raw.Length;
                var skel = HarakatLexicon.StripHarakat(raw).Normalize(System.Text.NormalizationForm.FormC);
                if (skel.Length == 0 || HarakatLexicon.HARAKAT.IsMatch(raw) || lexicon.ContainsKey(skel))
                {
                    @out.Add(raw);
                    continue;
                }
                var cps = Js.CodePoints(skel);
                var labels = await Predict(langTokIdx.Value, cps.Select(c => _meta.Chars.GetValueOrDefault(c, _unk)).ToList())
                    .ConfigureAwait(false);
                @out.Add(string.Concat(cps.Select((c, k) => c + HarOf(k < labels.Count ? labels[k] : "0"))));
            }
            @out.Add(text[cursor..]);
            return string.Concat(@out);
        }
    }

    /** Load a rider diacritizer from ONNX model bytes + sidecar meta. Session created once and reused. */
    public static async Task<IRiderDiacritizer> LoadRiderDiacritizer(byte[] modelBytes, RiderDiacritizerMeta meta)
    {
        var ort = await Onnx.LoadOrt("Rider neural diacritization").ConfigureAwait(false);
        var session = await ort.CreateInferenceSession(modelBytes).ConfigureAwait(false);
        return new Loaded(ort, session, meta);
    }

    /** Load the rider diacritizer from the model + meta beside this file. ⚠ DEGRADES TO NULL on ANY
     *  unavailability — a missing model or meta, or a failed ONNX runtime load — rather than throwing;
     *  callers fall back to the lexicon+default path. */
    public static async Task<IRiderDiacritizer?> CreateRiderDiacritizer()
    {
        byte[] bytes;
        RiderDiacritizerMeta meta;
        try
        {
            bytes = File.ReadAllBytes(DataPath.ResolveAllowMissing("core/riderDiacritizer.onnx"));
            meta = JsonSerializer.Deserialize<RiderDiacritizerMeta>(
                File.ReadAllText(DataPath.ResolveAllowMissing("core/riderDiacritizer.meta.json")), Jsonc.JsonOpts)!;
        }
        catch { return null; } // model or sidecar meta absent/corrupt
        try
        {
            return await LoadRiderDiacritizer(bytes, meta).ConfigureAwait(false);
        }
        catch { return null; } // onnxruntime-node absent or the session failed to build → sync fallback
    }
}

File: src/core/riderDiacritizer.ts
/**
 * Neural GENERALIZATION tier for the Perso-Arabic riders (Urdu, Persian, Pashto, Punjabi-Shahmukhi) — a shared
 * multilingual BiLSTM that restores short-vowel harakat on a bare skeleton, run via ONNX Runtime as an ASYNC
 * PRE-PASS. It fills the words the exact-match lexicon (harakatLexicon.ts) misses; a word the lexicon already
 * covers is LEFT BARE here so the (authoritative, gold) sync lexicon layer vocalizes it — giving the precedence
 * lexicon → neural → default. Vocalized output feeds the SYNC g2p, so phonemize() stays sync and dependency-free;
 * only this pre-pass touches ONNX. `onnxruntime-node` is an OPTIONAL dependency, imported lazily; the model is a
 * per-word restorer (one language TOKEN prepended, char sequence, argmax harakat per position). The int8 model
 * ships in-repo (like the Arabic diacritizer) but is optional at RUNTIME: if it — or `onnxruntime-node` — is
 * absent, the pre-pass degrades to a no-op and callers get the lexicon+default path. See
 * and tools/perso-arabic/export_onnx.py.
 */
import { readData, readDataText } from "./dataSource.ts";

import { HARAKAT, HARAKAT_G, stripHarakat } from "./harakatLexicon.ts";
import { loadOrt } from "./onnx.ts";
import { dataDir } from "./dataPath.ts";

// label → the combining harakat to append after a base letter. Mirrors the training VOWELS map (invert_harakat.ts /
// train_multilingual_harakat.py): a fatḥa, u damma, i kasra, o sukūn, F/N/K tanwīn, ^ dagger-alif (U+0670); a "~"
// prefix = shadda (gemination) + that vowel; "0" = bare (the model's default).
const HAR: Record<string, string> = {
    "0": "", a: "َ", u: "ُ", i: "ِ", o: "ْ",
    F: "ً", N: "ٌ", K: "ٍ", "^": "ٰ",
};
function harOf(label: string): string {
    if (label.startsWith("~")) return "ّ" + (HAR[label.slice(1)] ?? "");
    return HAR[label] ?? "";
}

// A Perso-Arabic word run: letters/joiners in U+0600–06FF + U+0750–077F, PLUS the ZWNJ/ZWJ (U+200C/200D) that
// glue morphemes in Persian/Urdu compounds — the model was trained on those as ONE sequence (a ZWNJ-split would
// change the LSTM context and the predicted harakat), MINUS the harakat (handled separately).
const WORD = /[؀-ۿݐ-ݿ‌‍]+/gu;  // ZWNJ, ZWJ

export interface RiderDiacritizerMeta {
    chars: Record<string, number>;
    labels: Record<string, number>;
    lang_tokens: Record<string, string>;
}
export interface RiderDiacritizer {
    /** Vocalize the words of `text` for `lang`, skipping any word whose skeleton is in `lexicon` (left bare for the
     *  sync lexicon layer). Position-preserving: only harakat are inserted; spacing/digits/punctuation are kept. */
    diacritize(text: string, lang: string, lexicon: ReadonlyMap<string, string>): Promise<string>;
}

/** Load a rider diacritizer from ONNX model bytes + sidecar meta. Session created once and reused. */
export async function loadRiderDiacritizer(modelBytes: Uint8Array, meta: RiderDiacritizerMeta): Promise<RiderDiacritizer> {
    const ort = await loadOrt("Rider neural diacritization");
    const session = await ort.InferenceSession.create(modelBytes);
    const ilabels: string[] = [];
    for (const [lab, idx] of Object.entries(meta.labels)) ilabels[idx] = lab;
    const UNK = meta.chars["<unk>"] ?? 1;
    const nLabels = ilabels.length;

    /** One skeleton → predicted harakat labels (per base char; the prepended lang-token position is dropped). */
    async function predict(langTokIdx: number, chars: number[]): Promise<string[]> {
        const seq = [langTokIdx, ...chars];
        const input = new ort.Tensor("int64", BigInt64Array.from(seq, BigInt), [1, seq.length]);
        const out = await session.run({ input });
        const logitsTensor = out["logits"] ?? out[Object.keys(out)[0]!];
        if (!logitsTensor) throw new Error("rider diacritizer ONNX produced no output");
        const logits = logitsTensor.data as Float32Array;
        const labels: string[] = [];
        for (let p = 1; p < seq.length; p++) { // p=0 is the lang token → skipped
            let best = 0, bestv = -Infinity;
            const base = p * nLabels;
            for (let o = 0; o < nLabels; o++) { const v = logits[base + o]!; if (v > bestv) { bestv = v; best = o; } }
            labels.push(ilabels[best]!);
        }
        return labels;
    }

    return {
        async diacritize(text: string, lang: string, lexicon: ReadonlyMap<string, string>): Promise<string> {
            const langTok = meta.lang_tokens[lang];
            const langTokIdx = langTok !== undefined ? meta.chars[langTok] : undefined;
            if (langTokIdx === undefined) return text; // unknown language → no-op
            // Rebuild the text with a single cursor (append the gap before each word, then the word's replacement),
            // avoiding shift bookkeeping and repeated full-string slicing.
            const out: string[] = [];
            let cursor = 0;
            for (const m of text.matchAll(WORD)) {
                const raw = m[0]!;
                out.push(text.slice(cursor, m.index!));
                cursor = m.index! + raw.length;
                // Respect writer-supplied harakat (like restoreHarakat) and leave lexicon-covered words BARE for the
                // authoritative sync lexicon layer — both are emitted unchanged.
                const skel = stripHarakat(raw).normalize("NFC");
                if (skel.length === 0 || HARAKAT.test(raw) || lexicon.has(skel)) { out.push(raw); continue; }
                const cps = [...skel];
                const labels = await predict(langTokIdx, cps.map((c) => meta.chars[c] ?? UNK));
                out.push(cps.map((c, k) => c + harOf(labels[k] ?? "0")).join(""));
            }
            out.push(text.slice(cursor));
            return out.join("");
        },
    };
}

/** Load the rider diacritizer from the model + meta beside this file. The model ships in-repo but this DEGRADES to
 *  undefined (callers fall back to the lexicon+default path) on ANY unavailability — a missing model or meta, OR a
 *  failed `onnxruntime-node` load (optional native dep absent / ABI mismatch) — rather than throwing. Call
 *  loadRiderDiacritizer directly if you want the underlying error surfaced. */
export async function createRiderDiacritizer(): Promise<RiderDiacritizer | undefined> {
    const dir = dataDir(import.meta.url);
    let bytes: Uint8Array, meta: RiderDiacritizerMeta;
    try {
        bytes = readData(`${dir}/riderDiacritizer.onnx`);
        meta = JSON.parse(readDataText(`${dir}/riderDiacritizer.meta.json`)) as RiderDiacritizerMeta;
    } catch { return undefined; } // model or sidecar meta absent/corrupt
    try {
        return await loadRiderDiacritizer(bytes, meta);
    } catch { return undefined; } // onnxruntime-node absent or the session failed to build → sync fallback
}
